### PR TITLE
Execute parameters

### PR DESCRIPTION
**Pull request by adespawn**
Allow to execute queries to database while passing parameters. Adds support for adding and receiving empty values (partially solves #79).

Changed the behavior of the driver, so it executes all queries as prepared. The goal of such change was to enable ability to correctly cast the JS objects into rust CqlValue. As discussed in person, this may not be the best approach, but the goal of this PR is to have an initial working version.

Updates rust driver version to 0.15.0.

Add simple implementation of batch queries through rust driver. Batch query object is kept in a new wrapper, exposed to JS driver.

Currently the database query works as follows:
1. Rust driver is called to prepare the query. 
2. JS driver takes the array of expected types from prepared query
3. JS driver parses provided parameters based on provided expected type
4. Rust driver is called with prepared query and parsed parameters to execute the query

Enable examples and integration tests that work with the new logic

### CODE REVIEW

File: src/utils.rs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+use std::fmt::Display;
+
+use napi::{Error, Status};
+
+pub fn js_generic_err(msg: impl Display) -> Error<Status> {
+  Error::new(Status::GenericFailure, msg)
+}
```
Comment (wprzytula) [inline src/utils.rs:7]: I'm happy to see an instant use of new knowledge from today's lab - `impl Display` :wink:

File: src/result.rs [hunk 2]
```diff
@@ -51,6 +54,7 @@ pub enum CqlType {
     Tuple,
     Uuid,
     Varint,
+    Custom,
```
Comment (PiotrJunior) [inline src/result.rs:57]: What is Custom type? It is not specified in rust driver.

File: lib/client.js [hunk 3]
```diff
@@ -792,62 +793,46 @@ class Client extends events.EventEmitter {
...
-            // Only Cassandra 2.1 and above supports named parameters
-            throw new errors.ArgumentError(
-                "Named parameters for simple statements are not supported, use prepare flag",
-            );
+            throw new Error(`TODO: Implement any support for named parameters`);
+            // // Only Cassandra 2.1 and above supports named parameters
+            // throw new errors.ArgumentError(
+            //   "Named parameters for simple statements are not supported, use prepare flag",
+            // );
         }
```
Comment (PiotrJunior) [inline lib/client.js:815]: Maybe add issue for every temporally disabled feature to not forget about fixing those.

File: lib/client.js [hunk 4]
```diff
@@ -792,62 +793,46 @@ class Client extends events.EventEmitter {
...
 
-        let request;
-
         if (!this.connected) {
-            // Micro optimization to avoid an async execution for a simple check
+            // Check this logic and decide if it's needed. Probably toDo while implementing (better) connection
+            // // Micro optimization to avoid an async execution for a simple check
```
Comment (PiotrJunior) [inline lib/client.js:819]: double `// //`

File: lib/client.js [hunk 5]
```diff
@@ -792,62 +793,46 @@ class Client extends events.EventEmitter {
...
-                (types.protocolVersion.supportsKeyspaceInRequest(version) &&
-                    execOptions.getKeyspace()) ||
-                this.keyspace;
-    
-        const { queryId, meta } = await PrepareHandler.getPrepared(this, lbp, query, queryKeyspace);
-        request = await this.#createExecuteRequest(query, queryId, execOptions, params, meta);
-        */
-        }
-
-        return await RequestHandler.send(request, execOptions, this);
```
Comment (PiotrJunior) [inline lib/client.js:850]: Also add this as an issue

File: lib/client.js (hunk 4 above)
Comment (adespawn) [reply lib/client.js:819]: That's intentional

File: lib/client.js [hunk 6]
```diff
@@ -792,62 +793,46 @@ class Client extends events.EventEmitter {
...
+        return new Promise(async (resolve, reject) => {
+            try {
+                let statement = await this.rust_client.prepareStatement(query);
+                let parsed_params = parseParams(statement, params);
```
Comment (ZuzaOsa) [inline lib/client.js:826]: Use camelcase in JS.

**Review by ZuzaOsa — APPROVED** [review]
One variable name comment, otherwise lgtm.

File: lib/client.js [hunk 7]
```diff
@@ -792,62 +793,46 @@ class Client extends events.EventEmitter {
     }
 
     /**
-     * Connects and handles the execution of prepared and simple statements.
+     * Wrapper for executing queries by rust driver
```
Comment (Stapox35) [inline lib/client.js:796]: Please check the rest of the private methods. They do not have arguments listed in JSDoc.

**Review by Stapox35 — CHANGES_REQUESTED** [review]
At the end of the `client.js` file is a mysterious description of a function that is not there....

**Comment by wprzytula** [comment]
> Changed the behavior of the driver, so it executes all queries as prepared.

Can we mimic the Rust driver's behaviour that all queries with nonempty parameter list are prepared, but those without parameters can be sent as unprepared? This should not cause any problems around serialization, because there is nothing to be serialized in such case.

File: src/request.rs [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+    #[napi]
+    pub fn from_text(val: String) -> QueryParameterWrapper {
+        QueryParameterWrapper {
+            parameter: CqlValue::Ascii(val),
+        }
+    }
+
+    #[napi]
+    pub fn from_set(val: Vec<&QueryParameterWrapper>) -> QueryParameterWrapper {
```
Comment (wprzytula) [inline src/request.rs:102]: This seems a very weird type for a set to me. Why have a `Vec` of references at all?
Perhaps this function should accept `impl Iterator<Item = &QueryParameterWrapper>`?

Comment (wprzytula) [reply src/request.rs:102]: Also, we don't seem to have list nor map supported here, only set. Why so?

File: src/request.rs [hunk 9]
```diff
@@ -1,17 +1,26 @@
...
 #[napi]
 pub struct QueryParameterWrapper {
-    parameter: CqlValue,
+    pub(crate) parameter: CqlValue,
+}
+
+#[napi]
+pub struct PreparedStatementWrapper {
+    pub(crate) query: PreparedStatement,
 }
```
Comment (wprzytula) [inline src/request.rs]: When introducing public structs like these, let's try to write at least minimal docstrings, so that reading the code is easier. Just by hovering the type's name we will get a short info what it is for.

File: src/session.rs [hunk 10]
```diff
@@ -9,7 +13,7 @@ pub struct SessionOptions {
 
 #[napi]
 pub struct SessionWrapper {
-    internal_session: Session,
+    internal: Session,
 }
```
Comment (wprzytula) [inline src/session.rs:17]: Such a rename should be done in a separate commit, not to clutter the commit that contains meaningful new logic.

File: src/request.rs [hunk 11]
```diff
@@ -1,17 +1,26 @@
...
 
 #[napi]
 pub struct QueryParameterWrapper {
-    parameter: CqlValue,
+    pub(crate) parameter: CqlValue,
+}
+
+#[napi]
+pub struct PreparedStatementWrapper {
+    pub(crate) query: PreparedStatement,
```
Comment (wprzytula) [inline src/request.rs:23]: Let's have this field named `prepared`, because `query` is unfortunately a usual name for the unprepared statement in the Rust driver.

File: src/session.rs [hunk 12]
```diff
@@ -31,18 +35,45 @@ impl SessionWrapper {
             .build()
             .await
             .map_err(err_to_napi)?;
-        Ok(SessionWrapper {
-            internal_session: sb,
-        })
+        Ok(SessionWrapper { internal: sb })
```
Comment (wprzytula) [inline src/session.rs:38]: The variable name should be `s`, not `sb`, because it's an already built Session, not SessionBuilder.

File: src/session.rs [hunk 13]
```diff
@@ -31,18 +35,45 @@ impl SessionWrapper {
...
     }
 
     #[napi]
     pub async fn query_unpaged_no_values(&self, query: String) -> napi::Result<QueryResultWrapper> {
         let query = self
-            .internal_session
+            .internal
```
Comment (wprzytula) [inline src/session.rs:44]: Naming nit: `query` -> `query_result`

File: src/session.rs [hunk 14]
```diff
@@ -31,18 +35,45 @@ impl SessionWrapper {
...
+            query: self
+                .internal
+                .prepare(statement)
+                .await
+                .map_err(err_to_napi)?,
+        })
+    }
+
+    #[napi]
+    pub async fn query_simple(
```
Comment (wprzytula) [inline src/session.rs:66]: Please explain the name.

Comment (wprzytula) [inline src/session.rs:66]: Please provide docstrings to these functions, as they are crucial for the new driver.

File: src/request.rs (hunk 8 above)
Comment (adespawn) [reply src/request.rs:102]: 1. This functions is exposed to napi, so I don't think iterators will work here.

2. Those types are implemented here: https://github.com/scylladb-zpp-2024-javascript-driver/scylladb-javascript-driver/pull/95

File: src/request.rs [hunk 15]
```diff
@@ -63,6 +63,13 @@ impl QueryParameterWrapper {
         }
     }
 
+    #[napi]
+    pub fn from_duration(val: &DurationWrapper) -> QueryParameterWrapper {
+        QueryParameterWrapper {
+            parameter: CqlValue::Duration(val.get_cql_duration()),
+        }
+    }
+
```
Comment (wprzytula) [inline src/request.rs:72]: This should belong to the previous commit.

File: lib/types/cql-utils.js [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+            return rust.QueryParameterWrapper.fromBlob(value);
+        case rust.CqlType.Boolean:
+            return rust.QueryParameterWrapper.fromBoolean(value);
+        case rust.CqlType.Counter:
+            return rust.QueryParameterWrapper.fromCounter(value);
+        case rust.CqlType.Double:
+            return rust.QueryParameterWrapper.fromDouble(value);
+        case rust.CqlType.Duration:
+            if (!(value instanceof Duration)) throw new TypeError("");
+            return rust.QueryParameterWrapper.fromDuration(value.getInternal());
```
Comment (wprzytula) [inline lib/types/cql-utils.js:33]: Why is this check present only for Duration?

Comment (wprzytula) [reply lib/types/cql-utils.js:33]: Well, and for Uuid and TimeUuid.

File: lib/types/cql-utils.js [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+            if (!(value instanceof Uuid)) throw new TypeError("");
+            return rust.QueryParameterWrapper.fromUuid(value.getInternal());
+        case rust.CqlType.Timeuuid:
+            if (!(value instanceof TimeUuid)) throw new TypeError("");
+            return rust.QueryParameterWrapper.fromTimeUuid(value.getInternal());
+        default:
+            // Or not yet implemented type
+            throw new ReferenceError(
+                `[INTERNAL DRIVER ERROR] Unknown type: ${type}`,
+            );
```
Comment (wprzytula) [inline lib/types/cql-utils.js:56]: Should it really be a `ReferenceError`?

File: lib/client.js [hunk 18]
```diff
@@ -792,62 +793,46 @@ class Client extends events.EventEmitter {
...
+                let rusty = await this.rustClient.querySimple(
+                    statement,
+                    parsedParams,
+                );
+                resolve(new ResultSet(rusty));
+            } catch (e) {
+                reject(e);
+            }
+        });
     }
```
Comment (wprzytula) [inline lib/client.js:836]: I don't see the previous code use `reject` at all, and you do. Please explain it in a comment.

File: test/integration/supported/client-execute-tests.js [hunk 19]
```diff
@@ -217,8 +221,9 @@ describe("Client @SERVER_API", function () {
             ];
             // no hint
             insertSelectTest(client, table, columns, values, null, done);
-        });
-        vit(
+        }); */
+        // No support for parameter hints
+        /* vit(
```
Comment (wprzytula) [inline test/integration/supported/client-execute-tests.js:226]: I strongly reject such practice of commenting out test cases without having them documented clearly and marked with a tag for easy lookup.
Tests are here to reliably measure our progress; we mustn't decrease their accuracy.

File: lib/types/cql-utils.js [hunk 20]
```diff
@@ -14,6 +14,9 @@ const Duration = require("./duration");
  * @returns {rust.QueryParameterWrapper}
  */
 function getWrapped(type, value) {
+    if (value === null || value === types.unset) {
+        return rust.QueryParameterWrapper.fromEmpty();
+    }
```
Comment (wprzytula) [inline lib/types/cql-utils.js:19]: `CqlValue::Empty` is something totally different than the DB concept of `null`. Please don't mix them up.

File: src/result.rs [hunk 21]
```diff
@@ -129,11 +129,16 @@ impl QueryResultWrapper {
...
+    pub fn get_columns(&self) -> napi::Result<Vec<Option<CqlValueWrapper>>> {
+        let s: Vec<Option<CqlValueWrapper>> = self
             .internal
             .iter()
-            .filter_map(|f| f.as_ref().map(|r| CqlValueWrapper { inner: r.clone() }))
+            .map(|f| {
+                if f.is_none() {
+                    return None;
+                }
+                f.as_ref().map(|r| CqlValueWrapper { inner: r.clone() })
```
Comment (wprzytula) [inline src/result.rs:140]: FP: let's use `Option::map` instead of imperative style.

**Review by wprzytula — CHANGES_REQUESTED** [review]

File: lib/types/cql-utils.js (hunk 16 above)
Comment (adespawn) [reply lib/types/cql-utils.js:33]: For those types we call internal function on the passed object. For other types, the program will throw an error with meaningful description, while it's not the case for those types.

(I've updated the error messages for those TypeErrors)

File: test/integration/supported/client-execute-tests.js (hunk 19 above)
Comment (adespawn) [reply test/integration/supported/client-execute-tests.js:226]: How would you suggest to mark those tests that are currently failing due to lack of nesesery features in the driver then?

Comment (wprzytula) [reply test/integration/supported/client-execute-tests.js:226]: At least with a `TODO:`. But it's best to come up with a dedicated keyword, e.g. `TODO: TESTING:` or something like this.

File: lib/client.js (hunk 18 above)
Comment (wprzytula) [reply lib/client.js]: It's simply me not knowing JS. Nvm.

File: lib/types/cql-utils.js (hunk 16 above)
Comment (wprzytula) [reply lib/types/cql-utils.js]: Could you put the above explanation in a comment?

File: src/request.rs [hunk 22]
```diff
@@ -123,3 +134,15 @@ impl QueryParameterWrapper {
         }
     }
 }
+
+#[napi]
+impl PreparedStatementWrapper {
+    #[napi]
+    pub fn get_expected_types(&self) -> Vec<CqlType> {
```
Comment (wprzytula) [inline src/request.rs:141]: Missing docstring.

File: src/session.rs [hunk 23]
```diff
@@ -52,6 +53,46 @@ impl SessionWrapper {
...
+    /// Creating Prepared statement may help to determine required types
+    ///
+    /// Currently `execute_unpaged` from rust driver is used, so no pagin is done
+    /// and there is no support for any query parameters
+    #[napi]
+    pub async fn execute_prepared(
+        &self,
+        query: &PreparedStatementWrapper,
+        params: Vec<&QueryParameterWrapper>,
+    ) -> napi::Result<QueryResultWrapper> {
```
Comment (wprzytula) [inline src/session.rs]: > there is no support for any query parameters

Really? Isn't that what this PR is for?

File: lib/types/cql-utils.js [hunk 24]
```diff
@@ -14,6 +14,9 @@ const Duration = require("./duration");
  * @returns {rust.QueryParameterWrapper}
  */
 function getWrapped(type, value) {
+    if (value === null || value === types.unset) {
+        return null;
+    }
```
Comment (wprzytula) [inline lib/types/cql-utils.js:19]: This commit still mentions "empty" values, whereas it's about null values.

File: test/integration/supported/client-execute-tests.js [hunk 25]
```diff
@@ -600,8 +600,7 @@ describe("Client @SERVER_API", function () {
                 done,
             );
         }); */
-        // issues #79
-        /* vit("2.2", "should accept unset as a valid value", function (done) {
+        vit("2.2", "should accept unset as a valid value", function (done) {
```
Comment (wprzytula) [inline test/integration/supported/client-execute-tests.js:603]: This test mentions `unset`, not `null`s. Isn't it contrary to the commit's description?

File: src/session.rs (hunk 23 above)
Comment (adespawn) [reply src/session.rs]: My mistake. I meant query option

File: test/integration/supported/client-execute-tests.js (hunk 25 above)
Comment (adespawn) [reply test/integration/supported/client-execute-tests.js:603]: The commit should enable both things

File: src/session.rs [hunk 26]
```diff
@@ -52,6 +61,73 @@ impl SessionWrapper {
...
+    }
+
+    /// Query a database with a given prepared query and provided parameters.
+    ///
+    /// Returns a wrapper of the value provided by the rust driver
+    ///
+    /// All parameters need to be wrapped into QueryParameterWrapper keeping CqlValue of correct type
+    /// Creating Prepared statement may help to determine required types
+    ///
+    /// Currently `execute_unpaged` from rust driver is used, so no pagin is done
```
Comment (wprzytula) [inline src/session.rs:88]: :recycle: typo: pagin -> paging

File: src/session.rs [hunk 27]
```diff
@@ -52,6 +61,73 @@ impl SessionWrapper {
...
+        Ok(PreparedStatementWrapper {
+            prepared: self
+                .internal
+                .prepare(statement)
+                .await
+                .map_err(err_to_napi)?,
+        })
+    }
+
+    /// Query a database with a given prepared query and provided parameters.
```
Comment (wprzytula) [inline src/session.rs:81]: :recycle: naming: `prepared query` -> `prepared statement`

File: src/request.rs [hunk 28]
```diff
@@ -0,0 +1,160 @@
...
+            parameter: CqlValue::Duration(val.get_cql_duration()),
+        }
+    }
+
+    #[napi]
+    pub fn from_float(val: f64) -> QueryParameterWrapper {
+        QueryParameterWrapper {
+            parameter: CqlValue::Float(val as f32),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (adespawn) [reply src/request.rs:80]: I think this is the best thing we can do here. We can't assume that value can be converted without loses even for small values like ``0.1``. By casting we just loses precision and we can't reasonably require JS part to provide value in a way that will be lossless.
``f32::try_from<f64>`` is not implemented.

File: src/result.rs [hunk 29]
```diff
@@ -70,71 +78,91 @@ pub enum CqlType {
...
         };
 
-        Some(
-            rows.iter()
-                .map(|f| RowWrapper {
-                    internal: f.columns.clone(),
-                })
-                .collect(),
-        )
+        let rows = r2.rows::<Row>().map_err(err_to_napi)?.flatten();
```
[the remark this answers is not in the document]
Comment (adespawn) [reply src/result.rs:97]: 2. I would prefer to do it later, as this would require investigation how old driver behaves when it encounters that kind of problems.

File: src/result.rs [hunk 30]
```diff
@@ -70,71 +78,91 @@ pub enum CqlType {
...
+        Ok(Some(
+            rows.map(|f| RowWrapper {
+                internal: f.columns,
+            })
+            .collect(),
+        ))
```
[the remark this answers is not in the document]
Comment (adespawn) [reply src/result.rs:103]: I don't understand the difference. Additionally ``collect`` after this change requires type annotation

File: lib/client.js [hunk 31]
```diff
@@ -608,45 +609,32 @@ class Client extends events.EventEmitter {
...
+                );
+                parametersRows.push([]);
+            } else {
+                let query = queries[i].query;
+                let params = queries[i].params;
+                let prepared = await this.rustClient.prepareStatement(query);
+                let parameters = parseParams(prepared, params);
+                preparedQueries.push(prepared);
+                parametersRows.push(parameters);
             }
```
[the remark this answers is not in the document]
Comment (adespawn) [reply lib/client.js:629]: This logic is taken from Datastax code. I'm not aware of any sensible way of doing what you asked for. Either way if some invalid value is provided, Error will be thrown.

File: src/result.rs (hunk 29 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply src/result.rs]: > I would prefer to do it later, as this would require investigation how old driver behaves when it encounters that kind of problems.

I insist on not allowing any errors to be silently ignored in our merged code.

File: lib/client.js (hunk 31 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply lib/client.js:629]: OK, I'm not going to wrestle with JS.

File: src/result.rs (hunk 29 above)
[the remark this answers is not in the document]
Comment (adespawn) [reply src/result.rs]: In this case I updated the code to panic in case on Deserialization error. The correct error handling I would still prefer to do at a later time.

[the remark this answers is not in the document]
Comment (wprzytula) [reply src/result.rs]: OK, I can accept this as a temporary measure.

**Comment by wprzytula** [comment]
:recycle: I suggest that bumping up the Rust driver's version be done in a separate PR, to reduce size of this PR (and future PRs, too).

File: lib/types/cql-utils.js [hunk 32]
```diff
@@ -0,0 +1,93 @@
...
+ * @param {CqlType} type
+ * @param {*} value
+ * @returns {rust.QueryParameterWrapper}
+ */
+function getWrapped(type, value) {
+    if (value === null || value === types.unset) {
+        return null;
+    }
+    // To increase clarity of the error messages, if we call some function on value,
+    // type is checked explicitly. In other cases default Error will be thrown.
```
Comment (wprzytula) [inline lib/types/cql-utils.js:21]: :question: :wrench: It's not clear to me what scenario this describes.

Comment (wprzytula) [reply lib/types/cql-utils.js:21]: :question:  What function? What value?

File: src/result.rs [hunk 33]
```diff
@@ -70,71 +75,93 @@ pub enum CqlType {
 #[napi]
 impl QueryResultWrapper {
     pub fn from_query(internal: QueryResult) -> QueryResultWrapper {
-        QueryResultWrapper { internal }
+        let value = match internal.clone().into_rows_result() {
+            Ok(v) => QueryResultVariant::RowsResult(v),
+            Err(_) => QueryResultVariant::EmptyResult(internal),
+        };
```
Comment (wprzytula) [inline src/result.rs:81]: :wrench: IIRC, `into_rows_result` may fail in two cases:
1. if the underlying Result response is not of Rows kind (correctly handled here)
2. if deserialization of Result metadata fails. In such case, the error should be propagated, not converted into `QueryResultVariant::EmptyResult`, because the current handling is basically silent ignoring of errors.

File: src/result.rs [hunk 34]
```diff
@@ -70,71 +75,93 @@ pub enum CqlType {
 #[napi]
 impl QueryResultWrapper {
     pub fn from_query(internal: QueryResult) -> QueryResultWrapper {
-        QueryResultWrapper { internal }
+        let value = match internal.clone().into_rows_result() {
```
Comment (wprzytula) [inline src/result.rs:78]: :wrench: Needless `clone()` - `internal` is an owned `QueryResult` that we can consume.

**Review by wprzytula — CHANGES_REQUESTED** [review]

**Comment by adespawn** [comment]
> ♻️ I suggest that bumping up the Rust driver's version be done in a separate PR, to reduce size of this PR (and future PRs, too).

It was part of  other PR that got merged into this one: https://github.com/scylladb-zpp-2024-javascript-driver/scylladb-javascript-driver/pull/111

File: lib/types/cql-utils.js (hunk 32 above)
Comment (adespawn) [reply lib/types/cql-utils.js]: I hope the comment is more meaningful now.

File: src/result.rs [hunk 35]
```diff
@@ -69,72 +77,97 @@ pub enum CqlType {
...
-        QueryResultWrapper { internal }
+    pub fn from_query(internal: QueryResult) -> napi::Result<QueryResultWrapper> {
+        let value = match internal.result_not_rows() {
+            Ok(_) => QueryResultVariant::EmptyResult(internal),
+            // We can unwrap, as error will be returned only if result_not_rows is OK
+            Err(_) => {
+                QueryResultVariant::RowsResult(internal.into_rows_result().map_err(err_to_napi)?)
+            }
+        };
+        Ok(QueryResultWrapper { internal: value })
```
Comment (wprzytula) [inline src/result.rs:88]: :wrench: This is logically correct, but not rusty (no clear exhaustive pattern matching). What should be done here is pattern matching on `internal.into_rows_result()`.

**Review by wprzytula — CHANGES_REQUESTED** [review]

File: src/result.rs (hunk 35 above)
Comment (adespawn) [reply src/result.rs:88]: Then we can't avoid cloning, as matching on it consumes it, and we can't create empty result from it

Comment (wprzytula) [reply src/result.rs:88]: [Have you looked at the error variants that can arise there?](https://docs.rs/scylla/latest/scylla/transport/query_result/enum.IntoRowsResultError.html)
`IntoRowsResultError::ResultNotRows(QueryResult)` allows for getting the `QueryResult` back.

**Review by wprzytula — APPROVED** [review]
